Adds protocol support to make-domain-requester

DarrenN opened the pull request:
When setting up a domain-requester you can pass the optional `#:protocol` keyword to enable HTTPS URLs. The default remains HTTP. 

Example:

```scheme
(make-domain-requester "httpbin.org" http-requester #:protocol "https")
```

DarrenN commented:
Hmmm, the tests pass but there seems to be an issue with coveralls?

jackfirth commented:
Rather than adding an option to `make-domain-requester`, how about adding a separate `make-https-requester` procedure that wraps a requester and transforms the urls it uses to https urls? That makes it easier to add https to requesters created via other means.

jackfirth commented:
Not sure why Coveralls is failing, and the Travis racket versions need to be updated anyways. I'll look into that next time I get a chance.

DarrenN commented:
@jackfirth that makes sense. Pushed up a new commit adding `make-https-requester`.

jackfirth commented on request/private/http-location.rkt:
```diff
@@ -7,18 +7,51 @@
+  (wrap-requester-location
+   (domain+relative-path->http-url "http" domain _) requester))
+
+(define (make-https-requester domain requester)
```
I think `make-https-requester` shouldn't work with a domain and should _only_ change the URL of requests made with the underlying requester to HTTPS. Something like `(wrap-requester-location http-url->https-url requester)`.

jackfirth commented on request/private/http-location.rkt:
```diff
@@ -7,18 +7,51 @@
+  (check-equal? "https" (url-scheme https-url))
+  
+  (check-pred requester? http-req)
+  (check-equal? 200 (http-response-code http-resp))
```
`check-equal?` uses the first argument as the actual value and the second as the expected one, so when these checks fail they'll report the failure with mismatched actual and expected values.

jackfirth commented on request/private/http-location.rkt:
```diff
@@ -7,18 +7,51 @@
+  (define https-req
+    (make-https-requester domain http-requester))
+ 
+  (define http-resp (get http-req "/ip"))
```
This test will fail in the Racket catalog builds because they're performed in an isolated VM without network access. I'd recommend changing this from `(module+ test ...)` to `(module+ integration-test ...)` and adding a line to the Travis script to run `raco test` on the integration test submodule.

DarrenN commented:
Okay I think this follows along with what you suggested:

``` scheme
(make-domain-requester domain (make-https-requester http-requester))
```

Also updated the tests and added the `integration-test` submodule. I _think_ I correctly modified the `.travis.yml` file as well.

jackfirth commented on .travis.yml:
```diff
@@ -15,5 +15,5 @@ before_install:
 
 script:
- - raco test $TRAVIS_BUILD_DIR
+ - raco test --submodule integration-test $TRAVIS_BUILD_DIR
```
You'll need _two_ test commands, the original one that runs unit tests and this one to run integration tests. Otherwise any (module+ test body ...) forms elsewhere in the package won't run. I haven't actually written any tests though, so technically not a big deal.

jackfirth commented:
This looks great :+1: Thanks for sticking with it through review, and thanks for implementing this! Would you be willing to write some docs for it? I'll try and fix the travis/coveralls issues tonight.

DarrenN commented:
This has been great, a nice lesson in composition for sure. Will write up some docs tonight. I've also been working on a `json-requester` branch, will review that some more in light of this discussion before I PR it.

Thank again!

jackfirth commented:
If you'd like I can assign #12 to you then, json requesters has been on my list for some time.

DarrenN commented:
Pushed up docs for `make-https-requester`. Sure I'll take a swing at #12

jackfirth commented:
Merged #23, fixed cover and now it builds on more versions. You'll have to rebase.

DarrenN commented:
Nice to see the tests run against so many different Rackets.
